File: src/main/java/com/devspace/conexfy/factories/ConConnectionAuthStrategyFactory.java
package com.devspace.conexfy.factories;

import java.util.Map;
import java.util.Optional;

import org.springframework.stereotype.Component;

import com.devspace.conexfy.enums.ConAuthTypeEnum;
import com.devspace.conexfy.strategies.auths.ConConnectionAuthStrategy;
import com.devspace.conexfy.strategies.auths.ConConnectionOAuth2Strategy;

@Component
public class ConConnectionAuthStrategyFactory {
    
    private final ConConnectionOAuth2Strategy conConnectionOAuth2Strategy;
    private final Map<ConAuthTypeEnum, ConConnectionAuthStrategy> strategies;

    public ConConnectionAuthStrategyFactory(ConConnectionOAuth2Strategy conConnectionOAuth2Strategy) {
        this.conConnectionOAuth2Strategy = conConnectionOAuth2Strategy;
        
        this.strategies = Map.of(
            ConAuthTypeEnum.OAUTH2, this.conConnectionOAuth2Strategy
        );
    }
   
    public Optional<ConConnectionAuthStrategy> getStrategy(ConAuthTypeEnum authType) {
        if (ConAuthTypeEnum.NO_AUTH == authType) return Optional.empty(); // No auth needed

        ConConnectionAuthStrategy strategy = strategies.get(authType);
        if (strategy != null) {
            return Optional.of(strategy);
        }
    
        throw new IllegalArgumentException("Unsupported auth type: " + authType);
    }
}

File: src/main/kotlin/com/devspace/conexfy/factories/ConConnectionAuthStrategyFactory.kt
package com.devspace.conexfy.factories

import com.devspace.conexfy.enums.ConAuthTypeEnum
import com.devspace.conexfy.strategies.auths.ConConnectionAuthStrategy
import com.devspace.conexfy.strategies.auths.ConConnectionOAuth2Strategy
import org.springframework.stereotype.Component
import java.util.*
import java.util.Map

/**
 * Factory for creating authentication strategies for connections.
 *
 * This factory provides a way to retrieve the appropriate authentication strategy
 * based on the specified authentication type.
 */
@Component
class ConConnectionAuthStrategyFactory(private val conConnectionOAuth2Strategy: ConConnectionOAuth2Strategy) {
    private val strategies: MutableMap<ConAuthTypeEnum, ConConnectionAuthStrategy> = Map.of<ConAuthTypeEnum, ConConnectionAuthStrategy>(
        ConAuthTypeEnum.OAUTH2, this.conConnectionOAuth2Strategy
    )

    fun getStrategy(authType: ConAuthTypeEnum): Optional<ConConnectionAuthStrategy> {
        if (ConAuthTypeEnum.NO_AUTH == authType) return Optional.empty<ConConnectionAuthStrategy>() // No auth needed

        val strategy = strategies[authType]
        if (strategy != null) {
            return Optional.of<ConConnectionAuthStrategy>(strategy)
        }

        throw IllegalArgumentException("Unsupported auth type: $authType")
    }
}
